Add GitExtensions.Extensibility

RussKie opened the pull request:
## Proposed changes

- ~The initial move of all shared contracts and helper API to https://github.com/gitextensions/gitextensions.extensibility (see https://github.com/gitextensions/gitextensions.extensibility/pull/31).~ Create a new project called Gitextensions.Extensibility and move of all shared contracts and helper API.
- Updated namespaces.

This change allows for a separation of concerns - shared contracts that affect the main app as well as Git Extensions plugins will now reside in Gitextensions.Extensibility project, and concrete implementations of the contracts reside in the app.
This is a stepping stone in a direction of complete separation of the contracts and future move of those to https://github.com/gitextensions/gitextensions.extensibility project.

I expect further work will be required to clean the contract declarations (e.g., removal of default args as those play tricks on developers) as well as a holistic reassessment of the API surface.


## Merge strategy

<!-- Change the following if the merge strategy should be changed:
- Squash merge (maintainer to decide merge message, PR submitter should cleanup commits/messages at PR approval).
- Rebase merge (PR submitter must change the commit message for the last commit).
- Merge commit. (PR submitter to rebase and squash before merges).
- To be decided later.
The maintainer may still request the contributor to squash and rebase, to make sure that merges and commit messages are clarified.
-->

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented:
There must be a very huge benefit of moving all these interfaces and utilities to a submodule / NuGet package.
Could you elaborate the benefits, please? Where do you draw the line between things exposed to plugins and the internal, more changeable implementation. 
They need to compensate for the friction added by switching to/from submodules and by split PRs.

gerhardol commented:
This will make development of the core app annoying, will slow down the app for the benefit of a few plugins. Most moved functionality does not make sense in a module, for instance GitModule have mostly app internal functions.
This will lead to duplicated code.

mstv commented:
It should not be too difficult to directly publish the GE Extensibility NuGet from the GE main repo, shouldn't it?

gerhardol commented:
With this, you know if the plugins require changes after a GE update. On the other hand, there are so much core functionality in Extensibility that the plugins will always have to be updated.
There are easier ways to achieve the strict compatibility without limiting the core app development.
Of course,  the plugin functionality could be limited (more like other interfaces?), but it would not be better for plugins.

Adding plugins to an existing app without a designed interface is not smooth.

RussKie commented:
Apologies for the radio silence; the end of last year was very demanding
and draining (mentally and physically). Right now I'm on a vacation with a
very limited access to a computer.
I will respond to the feedback and questions when I'm back in about a
month.

On Mon, 8 Jan 2024, 9:58 pm Michael Seibt, ***@***.***> wrote:

> this is precisely why I want the complete segregation of the interfaces
> and implementations so that we have design discussions and come up with API
> that both age gracefully with time, provide a solid foundation for
> extensibility and provide coherent and meaningful contract.
>
> I fully support this! But I have concerns / objections to move them out of
> the repo. I still think my above comments are clear.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/11431#issuecomment-1881653751>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXUZO655E3N7W2CQ67LYNQ6V3AVCNFSM6AAAAABAKT3PPCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBRGY2TGNZVGE>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

RussKie commented:
I rebased the changeset and split it in a way that should make the review easy. You can safely skip over 598072e, c4c65da and f9c9244 as those were mechanical auto-fixer changes.

mstv reviewed (commented):
LGTM
I have fixed the first three comments myself.

RussKie commented on GitExtensions.Extensibility/Translations/Xliff/TranslationUtil.cs:
```diff
@@ -0,0 +1,446 @@
+            if (field.IsPublic && !field.IsInitOnly)
+            {
+                // if public AND modifiable (NOT readonly)
+                Trace.WriteLine(string.Format("Skip field {0}.{1} [{2}]", obj.GetType().Name, field.Name, field.GetValue(obj)), "Translation");
```
It could be a bad rebase, I wasn't sure which one to take and took the wrong one :)

RussKie commented on GitExtensions.Extensibility/Settings/NumberSetting.cs:
```diff
@@ -0,0 +1,182 @@
+    {
+        if (typeof(T) == typeof(int))
+        {
+            return new NumericUpDownBinding((this as NumberSetting<int>)!, CustomControl as NumericUpDown);
```
Because it doesn't work...
![image](https://github.com/gitextensions/gitextensions/assets/4403806/2e74652f-d33a-4462-a00b-24946f941689)

mstv replied:
> Because it doesn't work...

... due to not to be expected technical restrictions of the compiler. `(T)variable` ought to be equivalent to `variable as T ?? throw new InvalidCastException(...)`.

Not in this case but basically, the `!` operator can be added such easily so that I fear it is added too carelessly.
If its accuracy is not obvious from a few preceding lines, it should be commented.
(And `null!` will always be an ugly hack to me.)

RussKie replied:
> Not in this case but basically, the `!` operator can be added such easily so that I fear it is added too carelessly.
> If its accuracy is not obvious from a few preceding lines, it should be commented.


Trust me, I dislike `!` as much as you do for NRTA, and don't use it freely.
I'll add a comment.

> (And `null!` will always be an ugly hack to me.)

Aye, though it's often unavoidable when annotating legacy code.

RussKie commented:
Thank you @mstv
